File: guide_creator_flow/src/guide_creator_flow/presets.py
"""Content presets — declarative bundles that bias and constrain generation.

A preset here is the *genre* dimension (educational / animation / images) and is
distinct from the aspect ``preset`` in ``main.py`` (landscape / reel). Each
bundle both hard-constrains its defining trait — the ``media_policy`` is enforced
in code, before ``fetch_assets`` runs, so the genre's promise holds regardless of
what the crew chose — and softly steers tone via ``guidance`` snippets injected
into the crew and director prompts.

Empty / ``None`` fields mean "no opinion — fall through to the default
behaviour". Behaviour lives here, next to the vocabularies in ``ir_builder``; the
server mirrors only id/label/description for the composer UI. A user's custom
preset travels in the trigger payload as a full bundle of the same shape and is
merged onto ``auto`` and used verbatim.
"""

from __future__ import annotations

import copy

# ------------------------------------------------------------ built-in presets

PRESETS: dict[str, dict] = {
    "auto": {
        "id": "auto",
        "label": "Auto",
        "description": "Let the director choose everything from the prompt.",
        "media_policy": {"force": None, "prefer": None, "block": None, "ken_burns_all": False},
        "theme": {"mood_pool": None, "font_pool": None},
        "voice_default": "",
        "music_default": "",
        "guidance": {"writer": "", "director": "", "asset": "", "design": ""},
    },
    "educational": {
        "id": "educational",
        "label": "Educational",
        "description": "Clear explainers driven by diagrams, charts and title cards.",
        "media_policy": {
            "force": None,
            "prefer": ["graphic"],
            "block": None,
            "ken_burns_all": False,
        },
        "theme": {
            "mood_pool": ["premium", "calm"],
            "font_pool": ["grotesque_sans", "geometric_sans", "classic_serif"],
        },
        "voice_default": "",
        "music_default": "",
        "guidance": {
            "writer": "Teach one idea per scene in plain, structured language; define terms "
                      "simply and build from the basics to a clear takeaway.",
            "director": "Favour data-driven motion graphics — diagrams, charts, counters, "
                        "annotated callouts and title cards — over stock footage. Keep every "
                        "layout clean and legible.",
            "asset": "Prefer 'graphic' for concepts, numbers, processes and comparisons; use a "
                     "photo only for a concrete real-world subject a diagram cannot show.",
            "design": "A calm, credible, editorial register with a highly legible typeface.",
        },
    },
    "animation": {
        "id": "animation",
        "label": "Animation",
        "description": "Playful animated illustrations and motion graphics — no stock footage.",
        "media_policy": {
            "force": ["lottie", "graphic"],
            "prefer": None,
            "block": None,
            "ken_burns_all": False,
        },
        "theme": {
            "mood_pool": ["playful", "bold"],
            "font_pool": ["geometric_sans", "display_heavy", "techno_sans"],
        },
        "voice_default": "",
        "music_default": "",
        "guidance": {
            "writer": "Keep an energetic, playful voice with vivid metaphors an animated "
                      "illustration can act out.",
            "director": "Tell the story entirely through animated vector illustrations and "
                        "motion graphics — characters, mascots, object metaphors and data "
                        "graphics. No live-action or photographic footage.",
            "asset": "Use 'lottie' for characters, mascots, object metaphors and actions; use "
                     "'graphic' for numbers, diagrams and title cards. Never 'photo' or 'video'.",
            "design": "A bold, playful register with an expressive display typeface.",
        },
    },
    "images": {
        "id": "images",
        "label": "Images",
        "description": "Cinematic still photography with Ken Burns motion.",
        "media_policy": {
            "force": ["photo"],
            "prefer": None,
            "block": None,
            "ken_burns_all": True,
        },
        "theme": {
            "mood_pool": ["premium", "nature", "calm"],
            "font_pool": ["editorial_serif", "classic_serif", "editorial_display"],
        },
        "voice_default": "",
        "music_default": "",
        "guidance": {
            "writer": "Write evocative, image-led narration where every beat calls to mind a "
                      "single strong photograph.",
            "director": "Tell the story through a sequence of striking still photographs with "
                        "slow Ken Burns motion; avoid diagrams and animation.",
            "asset": "Use 'photo' for every scene — concrete, photogenic real-world subjects "
                     "with strong composition. No graphics, lotties or video.",
            "design": "A premium, editorial register suited to full-bleed photography.",
        },
    },
}

DEFAULT_GENRE = "auto"


# --------------------------------------------------------------- bundle resolve

def resolve(genre: str, custom_bundle: dict | None = None) -> dict:
    """The active bundle for a run: a custom bundle (merged onto ``auto`` to fill
    any missing keys) when supplied, else the built-in for ``genre``, else auto."""
    base = copy.deepcopy(PRESETS["auto"])
    chosen = custom_bundle if custom_bundle else PRESETS.get(genre)
    if not chosen:
        return base
    return _merge(base, chosen)


def _merge(base: dict, over: dict) -> dict:
    out = copy.deepcopy(base)
    for key, value in (over or {}).items():
        if isinstance(value, dict) and isinstance(out.get(key), dict):
            out[key] = _merge(out[key], value)
        else:
            out[key] = value
    return out


# ------------------------------------------------------ hard media-policy pass

# Words in a curator query that signal a data/graphic-friendly scene, used to
# route a coerced entry to 'graphic' rather than an illustration.
_DATA_HINTS = (
    "chart", "graph", "percent", "%", "number", "count", "data", "growth",
    "stat", "trend", "compare", " vs ", "rate", "ratio", "timeline",
)

# Query prefixes that name a specific person/company face — stripped when a
# preset blocks them so the fallback treatment never fetches that face.
_NAME_PREFIXES = ("portrait of", "photo of", "logo of", "headshot of")


def _looks_like_data(query: str) -> bool:
    q = f" {(query or '').lower()} "
    if any(hint in q for hint in _DATA_HINTS):
        return True
    return any(ch.isdigit() for ch in q)


def _blocked(query: str, block: list[str]) -> bool:
    q = (query or "").strip().lower()
    for raw in block or []:
        b = str(raw).lower().strip()
        if b == "portrait" and (q.startswith(("portrait of", "photo of", "headshot of"))
                                or "headshot" in q):
            return True
        if b == "logo" and q.startswith("logo of"):
            return True
        if b and b in q:
            return True
    return False


def _neutralize(query: str) -> str:
    """Strip a person/company name out of a blocked query so the fallback
    treatment doesn't try to fetch that face ('portrait of Elon Musk' -> '')."""
    low = (query or "").strip().lower()
    for prefix in _NAME_PREFIXES:
        if low.startswith(prefix):
            return ""
    return (query or "").strip()


def _coerce(entry: dict, allowed: list[str]) -> None:
    """Rewrite ``entry['media_type']`` to the closest allowed type."""
    if entry.get("media_type") in allowed:
        return
    if "graphic" in allowed and _looks_like_data(entry.get("query", "")):
        entry["media_type"] = "graphic"
        return
    for candidate in allowed:            # first non-graphic option reads best…
        if candidate != "graphic":
            entry["media_type"] = candidate
            return
    entry["media_type"] = allowed[0]     # …else whatever is allowed


def apply_media_policy(asset_plan: list[dict], bundle: dict) -> list[dict]:
    """Hard-enforce a preset's media policy on the curator's plan, in place, and
    return the same list. Runs BEFORE ``fetch_assets`` so the genre's promise
    holds no matter what the LLM chose. 'upload' entries (the user's own footage)
    are never touched; ``prefer`` is intentionally soft (honoured via the asset
    guidance, not rewritten here) — only ``force`` and ``block`` are hard."""
    policy = (bundle or {}).get("media_policy") or {}
    force = policy.get("force") or None
    block = policy.get("block") or None
    if not force and not block:
        return asset_plan
    for entry in asset_plan:
        if not isinstance(entry, dict) or entry.get("media_type") == "upload":
            continue
        if block and _blocked(entry.get("query", ""), block):
            entry["query"] = _neutralize(entry.get("query", ""))
            _coerce(entry, force or (policy.get("prefer") or ["graphic"]))
            continue
        if force:
            _coerce(entry, force)
    return asset_plan

File: server/src/presets.js
import { randomUUID } from 'node:crypto';
import { admin } from './supabaseAdmin.js';
import { persistenceEnabled } from './config.js';
import { readIr } from './snapshot.js';

// Content presets (the "genre" dimension) — Educational / Animation / Images,
// plus user-created ones. Behaviour lives in the Python flow (presets.py); the
// server owns only the UI metadata for the built-ins and CRUD for custom
// presets. A custom preset is a full bundle (same shape as the Python built-ins)
// that rides the trigger payload as `preset_bundle` and is used verbatim.

// Built-in metadata for the composer picker. `id` must match Python PRESETS so
// the flow resolves behaviour from the id alone (no bundle needed).
export const BUILTIN_PRESETS = [
  { id: 'auto', label: 'Auto', description: 'Let the director choose everything from the prompt.', icon: '', builtin: true },
  { id: 'educational', label: 'Educational', description: 'Clear explainers driven by diagrams, charts and title cards.', icon: '', builtin: true },
  { id: 'animation', label: 'Animation', description: 'Playful animated illustrations and motion graphics — no stock footage.', icon: '', builtin: true },
  { id: 'footage', label: 'Footage', description: 'Real videos and photos only — no generated graphics.', icon: '', builtin: true },
  { id: 'images', label: 'Images', description: 'Cinematic still photography with Ken Burns motion.', icon: '', builtin: true },
];

const BUILTIN_IDS = new Set(BUILTIN_PRESETS.map((p) => p.id));

// In-memory store when persistence is off (mirrors store.js's `mem` pattern).
const mem = new Map(); // id -> { id, userId, label, bundle, createdAt }

const MEDIA_TYPES = ['photo', 'video', 'graphic', 'lottie'];

// Normalise an arbitrary client-supplied bundle into the canonical shape the
// flow expects. Unknown/garbage fields are dropped; missing ones stay null so
// the flow falls through to default behaviour.
export function normalizeBundle(input = {}, { id, label } = {}) {
  const mp = input.media_policy || {};
  const th = input.theme || {};
  const gd = input.guidance || {};
  const cleanList = (v, allowed) =>
    Array.isArray(v) && v.length
      ? v.map(String).filter((x) => !allowed || allowed.includes(x))
      : null;
  return {
    id: id || input.id || `custom-${randomUUID().slice(0, 8)}`,
    label: (label || input.label || 'My preset').slice(0, 60),
    description: (input.description || 'Custom preset.').slice(0, 200),
    media_policy: {
      force: cleanList(mp.force, MEDIA_TYPES),
      prefer: cleanList(mp.prefer, MEDIA_TYPES),
      block: cleanList(mp.block, null),
      ken_burns_all: Boolean(mp.ken_burns_all),
    },
    theme: {
      mood_pool: cleanList(th.mood_pool, null),
      font_pool: cleanList(th.font_pool, null),
    },
    voice_default: input.voice_default || '',
    music_default: input.music_default || '',
    guidance: {
      writer: (gd.writer || '').slice(0, 400),
      director: (gd.director || '').slice(0, 400),
      asset: (gd.asset || '').slice(0, 400),
      design: (gd.design || '').slice(0, 400),
    },
  };
}

// ---------------------------------------------------------------- read

// { builtins, custom } for the composer. Custom presets are the caller's own.
export async function listPresets(userId) {
  return { builtins: BUILTIN_PRESETS, custom: await listCustom(userId) };
}

async function listCustom(userId) {
  if (!persistenceEnabled) {
    return [...mem.values()].filter((p) => p.userId === userId).map(toView);
  }
  const { data, error } = await admin
    .from('presets').select('*').eq('user_id', userId).order('created_at', { ascending: false });
  if (error) { console.warn('[db] preset list failed:', error.message); return []; }
  return (data || []).map((r) => toView({ id: r.id, label: r.label, bundle: r.bundle, createdAt: r.created_at }));
}

function toView(p) {
  return { id: p.id, label: p.label, description: p.bundle?.description || '', builtin: false, bundle: p.bundle };
}

// The resolved bundle for a genre id: null for a built-in (the flow resolves it
// from the id) or an unknown id; the stored bundle for a custom preset.
export async function bundleFor(userId, genre) {
  if (!genre || BUILTIN_IDS.has(genre)) return null;
  if (!persistenceEnabled) {
    const p = mem.get(genre);
    return p && p.userId === userId ? p.bundle : null;
  }
  const { data, error } = await admin
    .from('presets').select('bundle').eq('user_id', userId).eq('id', genre).maybeSingle();
  if (error) { console.warn('[db] preset fetch failed:', error.message); return null; }
  return data?.bundle || null;
}

// ---------------------------------------------------------------- write

// Create a custom preset either from an explicit bundle (the composer form) or
// from an existing project's rendered look (`fromProjectId`, the editor's
// "save style as preset").
export async function createPreset(userId, { label, bundle, fromProjectId } = {}, resolveProject) {
  let raw = bundle;
  if (fromProjectId) {
    const project = resolveProject ? await resolveProject(fromProjectId) : null;
    if (!project) throw new Error('project not found');
    raw = await bundleFromProject(project);
  }
  const id = `custom-${randomUUID().slice(0, 8)}`;
  const norm = normalizeBundle(raw || {}, { id, label });
  const rec = { id, userId, label: norm.label, bundle: norm, createdAt: new Date().toISOString() };
  mem.set(id, rec);
  if (persistenceEnabled) {
    const { error } = await admin.from('presets').insert({
      id, user_id: userId, label: norm.label, bundle: norm,
    });
    if (error) console.warn('[db] preset insert failed:', error.message);
  }
  return toView(rec);
}

export async function deletePreset(userId, id) {
  const p = mem.get(id);
  if (p && p.userId === userId) mem.delete(id);
  if (persistenceEnabled) {
    const { error } = await admin.from('presets').delete().eq('user_id', userId).eq('id', id);
    if (error) { console.warn('[db] preset delete failed:', error.message); return false; }
  }
  return true;
}

// Derive a preset bundle from a finished project: copy its theme and infer the
// media policy from the actual layer-type mix of its rendered IR.
async function bundleFromProject(project) {
  const ir = await readIr(project.id).catch(() => null);
  const theme = ir?.metadata?.theme || {};
  const counts = { photo: 0, video: 0, graphic: 0, lottie: 0 };
  const scenes = ir?.scenes || ir?.timeline?.scenes || [];
  let total = 0;
  for (const scene of scenes) {
    for (const layer of scene.layers || []) {
      const t = layer.type === 'image' ? 'photo' : layer.type; // IR 'image' == asset 'photo'
      if (t in counts) { counts[t] += 1; total += 1; }
    }
  }
  const ranked = Object.entries(counts).filter(([, n]) => n > 0).sort((a, b) => b[1] - a[1]);
  const media_policy = { force: null, prefer: null, block: null, ken_burns_all: false };
  if (ranked.length) {
    const [top, topN] = ranked[0];
    if (topN / total >= 0.6) media_policy.force = [top];  // dominated → hard force
    else media_policy.prefer = [top];                     // mixed → soft prefer
    if (top === 'photo') media_policy.ken_burns_all = true;
  }
  return {
    label: deriveLabel(project),
    description: `Saved from “${project.title || project.prompt}”.`,
    media_policy,
    theme: {
      mood_pool: theme.mood ? [theme.mood] : null,
      font_pool: theme.font?.style ? [theme.font.style] : null,
    },
    voice_default: project.voice || '',
    music_default: project.music || '',
  };
}

function deriveLabel(project) {
  const base = (project.title || project.prompt || 'My style').trim();
  return `${base.split(/\s+/).slice(0, 4).join(' ')} style`.slice(0, 60);
}

export { BUILTIN_IDS };
